airbyte-integrations/connectors/destination-mongodb/src/main/java/io/airbyte/integrations/destination/mongodb/MongodbRecordConsumer.java

```diff
@@ -0,0 +1,164 @@
+import java.util.stream.StreamSupport;
+import org.bson.Document;
+import org.joda.time.LocalDateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MongodbRecordConsumer extends FailureTrackingAirbyteMessageConsumer implements AirbyteMessageConsumer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongodbRecordConsumer.class);
+
+  private static final String AIRBYTE_DATA = "_airbyte_data";
+  private static final String AIRBYTE_EMITTED_AT = "_airbyte_emitted_at";
+
+  private final Map<AirbyteStreamNameNamespacePair, MongodbWriteConfig> writeConfigs;
+  private final MongoDatabase mongoDatabase;
+  private final ConfiguredAirbyteCatalog catalog;
+  private final Consumer<AirbyteMessage> outputRecordCollector;
+  private final ObjectMapper objectMapper;
+
+  private AirbyteMessage lastStateMessage = null;
+
+  public MongodbRecordConsumer(Map<AirbyteStreamNameNamespacePair, MongodbWriteConfig> writeConfigs,
+                               MongoDatabase mongoDatabase,
+                               ConfiguredAirbyteCatalog catalog,
+                               Consumer<AirbyteMessage> outputRecordCollector) {
+    this.writeConfigs = writeConfigs;
+    this.mongoDatabase = mongoDatabase;
+    this.catalog = catalog;
+    this.outputRecordCollector = outputRecordCollector;
+    this.objectMapper = new ObjectMapper();
+  }
+
+  @Override
+  protected void startTracked() {
+    // todo (cgardens) - move contents of #write into this method.
+  }
+
+  @Override
+  protected void acceptTracked(AirbyteMessage message) {
+    if (message.getType() == AirbyteMessage.Type.STATE) {
+      lastStateMessage = message;
+    } else if (message.getType() == AirbyteMessage.Type.RECORD) {
+      final AirbyteRecordMessage recordMessage = message.getRecord();
+      AirbyteStreamNameNamespacePair pair = AirbyteStreamNameNamespacePair.fromRecordMessage(recordMessage);
+
+      if (!writeConfigs.containsKey(pair)) {
+        throw new IllegalArgumentException(
+            String.format("Message contained record from a stream that was not in the catalog. \ncatalog: %s , \nmessage: %s",
+                Jsons.serialize(catalog), Jsons.serialize(recordMessage)));
+      }
+      try {
+        Map<String, Object> result = objectMapper.convertValue(recordMessage.getData(), new TypeReference<>() {});
+        var newDocument = new Document();
+        newDocument.put(AIRBYTE_DATA, new Document(result));
+        newDocument.put(AIRBYTE_EMITTED_AT, new LocalDateTime().toString());
+
+        var writeConfig = writeConfigs.get(pair);
+        var collection = writeConfig.getCollection();
+
+        var documents = writeConfig.getDocuments();
+        if (!documents.contains(newDocument.get(AIRBYTE_DATA, Document.class))) {
+          collection.insertOne(newDocument);
+        }
+      } catch (RuntimeException e) {
+        LOGGER.error("Got an error while writing message:" + e.getMessage());
```

Comment:
Please add moe data in logs. As much as you can collect from here i.e. streams, catalogs, current msg that failed to process etc. We already faced a lot of cases when clients are not sufficient to understand where the issue is.  Some for other case.  As an example, you may check the destination-bigquery consumer. Thanks

Reply:
added new log with more data